### ISSUE TITLE
Xpetra: BlockedCrsMatrix::Merge fails if block matrices include too many entries

### ISSUE THREAD

**Issue by kliegeois** [trilinos/Trilinos#6765]
## Bug Report
@trilinos/xpetra 

### Description
Since the removal of the DynamicProfile option for the creation of CrsGraph, the static size of the graph must be known when calling the constructor and, if the number of entries per row given as an argument to the constructor is smaller than the number of entries that we actually try to insert into a given row of the graph, an error is thrown.

Such an error appends while merging (i.e. calling the member function Merge) BlockedCrsMatrix if one row has more than 33 entries:
```
Tpetra::CrsGraph<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::insertGlobalIndicesImpl: On MPI Process 0: Not enough capacity to insert 36 indices into local row 12, which currently has 0 entries and total allocation size 33.
```
this comes from line 1371 of https://github.com/trilinos/Trilinos/blob/develop/packages/xpetra/src/BlockedCrsMatrix/Xpetra_BlockedCrsMatrix.hpp:

```
RCP<Matrix> sparse = MatrixFactory::Build(getRangeMapExtractor()->getFullMap(), 33);
```

I suppose that the easiest way to resolve this issue is to replace that line by:
```
RCP<Matrix> sparse = MatrixFactory::Build(getRangeMapExtractor()->getFullMap(), this->getNodeMaxNumRowEntries());
```

I can create a PR with that modification and writing a test for that.

**Comment by mhoemmen** [comment]
@trilinos/tpetra

**Comment by mhoemmen** [comment]
Please don't just resort to the max number of entries.  This might use a LOT more memory.  Memory is already a problem; see e.g., #6663.  It's best to count the number of entries you need separately for each row.

**Comment by kliegeois** [comment]
Thanks for your comment.
I am looking into that; this will require a bit more work as some member functions such as `getNumEntriesInLocalRow` are not yet fully implemented for blockedCrsMatrix but it will be way better.
I should be able to create a PR next week with that.

**Comment by mhoemmen** [comment]
@kliegeois It's worth it, in my view.  We have no memory use testing for Trilinos, so we have to be really careful about changes that could result in a lot more memory use for problems with irregular structure (and thus are not like typical test problems).

**Comment by kliegeois** [comment]
@mhoemmen I am on it, it takes a bit of time as `getNumEntriesInLocalRow` are not yet fully implemented and Xpetra supports both Xpetra and Thyra mode.

**Comment by kliegeois** [comment]
Resolved by #6834 .
